File: services/analytics-service/src/modules/health/services/health.service.ts
import { Injectable, ServiceUnavailableException } from '@nestjs/common';
import { ErrorCode } from '../../../common/constants/error-code.enum';
import { RedisService } from '../../../common/utils/redis.service';
import { AnalyticsRepository } from '../../analytics/repositories/analytics.repository';

@Injectable()
export class HealthService {
  constructor(
    private readonly analyticsRepository: AnalyticsRepository,
    private readonly redisService: RedisService
  ) {}

  async getHealth(): Promise<Record<string, unknown>> {
    return {
      status: 'ok',
      service: 'analytics-service',
      timestamp: new Date().toISOString()
    };
  }

  async getReadiness(): Promise<Record<string, unknown>> {
    const postgres = await this.analyticsRepository.ping();
    const redis = this.redisService.isEnabled() ? await this.redisService.ping() : true;

    if (!postgres || !redis) {
      throw new ServiceUnavailableException({
        code: ErrorCode.SERVICE_UNAVAILABLE,
        message: 'Dependencies are not ready',
        details: {
          postgres,
          redis
        }
      });
    }

    return {
      status: 'ready',
      dependencies: {
        postgres,
        redis
      },
      timestamp: new Date().toISOString()
    };
  }

  async getLiveness(): Promise<Record<string, unknown>> {
    return {
      status: 'alive',
      service: 'analytics-service',
      timestamp: new Date().toISOString()
    };
  }
}

File: services/live-service/internal/service/health_service.go
package service

import (
	"context"

	"live-service/internal/repository"
)

type HealthService struct {
	appName string
	repo    repository.Repository
	redis   *RedisService
}

func NewHealthService(appName string, repo repository.Repository, redis *RedisService) *HealthService {
	return &HealthService{appName: appName, repo: repo, redis: redis}
}

func (s *HealthService) Health() map[string]any {
	return map[string]any{"service": s.appName, "status": "ok"}
}

func (s *HealthService) Live() map[string]any {
	return map[string]any{"service": s.appName, "status": "live"}
}

func (s *HealthService) Ready(ctx context.Context) (map[string]any, error) {
	if err := s.repo.Ping(ctx); err != nil {
		return nil, err
	}
	if s.redis != nil {
		if err := s.redis.Ping(ctx); err != nil {
			return nil, err
		}
	}
	return map[string]any{"service": s.appName, "status": "ready"}, nil
}
